[Step2] 수강신청(도메인 모델)

firewoody237 opened the pull request:
# Description
- 요구사항에 따라 Course 스키마를 변경했습니다.
- DB 구현은 하지않고, 순수한 도메인만 개발하였습니다.

# 느낀점
- 사실 이런식으로 DB와 관련없는 도메인 우선 개발은 처음인데, 다음 스텝에서 어떻게 적용 해 나아갈지 너무 궁금하네요 ㅎㅎ
- 설계에 이틀 가까이 너무 오랜 시간이 걸린 것 같습니다. 덕분에 개발은 하루만에 끝났지만 완성도가 어떨지 리뷰 받아봐야 알 것 같네요.
- 잘 부탁드립니다!!

firewoody237 committed:
docs: 구현 역할 문서 작성

firewoody237 committed:
feat: Cover 및 하위 도메인 구현

firewoody237 committed:
feat: Session 하위 도메인 구현

firewoody237 committed:
feat: Session 상태 구현

firewoody237 committed:
feat: SessionName 구현

firewoody237 committed:
feat: Session - FreeSession 구현

firewoody237 committed:
feat: 무료 강의를 위한 Capacity 수정

firewoody237 committed:
feat: 유료강의 기능 개발 및 그를 위한 도메인 조정

firewoody237 committed:
feat: course 기능 개발

firewoody237 committed:
feat: 테스트 오류로 인한 course 스키마 변경

firewoody237 committed:
refactor: Session내 Course 추가

firewoody237 committed:
docs: 설계도 변경

firewoody237 committed:
refactor: DeleteHistory 정적 생성자 사용

mskangg commented on README.md:
```diff
@@ -46,3 +46,31 @@
 - [x] Question 삭제 리스트 반환 구현
+
+# 수강 신청
+- 도메인 역할 도표
+![도메인 역할 도표](./images/image003.png)
```
👏 👏

mskangg commented on README.md:
```diff
@@ -46,3 +46,31 @@
+- 도메인 역할 도표
+![도메인 역할 도표](./images/image003.png)
+
+- [x] ImageExtensionType
```
이번엔 투두리스트는 없네요 🤔

firewoody237 replied:
아 이번에는 구현 리스트 자체는 `image003`이 충분히 설명하고 있다고 생각했고,
커밋을 진행하며 요구사항을 하나씩 지워가면서 진행했습니다.
그래서 큰 목록들(구현 기능)에 대한 Todo리스트만 작성했습니다..!

mskangg replied:
투두리스트는 작업하면서 빠뜨린게 없나 체크하는 용도도 있으니, 작성하는

mskangg commented on src/main/java/nextstep/courses/domain/Generation.java:
```diff
@@ -0,0 +1,14 @@
+
+    private final int generation;
+
+    public Generation(int generation) {
+        this.generation = generation;
+    }
```
기수는 음수는 존재하지 않으니, 음수체크정돈 하는게 좋을것 같아요

firewoody237 replied:
0기수 이하는 없다고 판단하여, 검증 추가하였습니다!

mskangg commented on src/main/java/nextstep/exception/CapacityException.java:
```diff
@@ -0,0 +1,8 @@
+package nextstep.exception;
+
+public class CapacityException extends UserDefinedException {
```
예외상황을 구체화하여 만드셨군요 👍

mskangg commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -79,6 +79,6 @@ public String toString() {
     public DeleteHistory delete() {
         this.setDeleted(true);
-        return new DeleteHistory(ContentType.ANSWER, this.id, this.writer, LocalDateTime.now());
+        return DeleteHistory.createAnswer(this.id, this.writer, LocalDateTime.now());
```
👍

mskangg commented on src/main/java/nextstep/session/domain/Capacity.java:
```diff
@@ -0,0 +1,51 @@
+        return new Capacity(maxCapacity);
+    }
+
+    public static Capacity create() {
+        return new Capacity(Integer.MAX_VALUE);
+    }
```
`무료 강의는 최대 수강 인원 제한이 없다.` 를 이렇게 푸셨군요 👍 
무료 강의를 생성할때 `Integer.MAX_VALUE` 를 넣어 생성하는것도 방법이 될것 같아요!

```java
// 무료강의.class

// 수강인원을 만들어줘(최대로)
Capacity.create(Integer.MAX_VALUE)
```

firewoody237 replied:
조금 고민해본 결과 무료이냐 아니냐에 대한 역할은 `Capacity`가 지고있을 필요는 없어보여
호출하는 `FreeSession`으로 넘겼습니다!

mskangg commented on src/main/java/nextstep/session/type/ImageExtensionType.java:
```diff
@@ -0,0 +1,29 @@
+        return Arrays.stream(ImageExtensionType.values())
+                .filter(type -> type.extension.equals(extension))
+                .findAny()
+                .orElseThrow(IllegalArgumentException::new);
```
이부분은 예외메세지를 세부구현하지 않은 이유가 있을까용? 🤔

firewoody237 replied:
리뷰해주시기 전까지 3단계를 임시로 구현해보고 있었는데,
저 부분에선 꼭 상태파악을 위한 메시지를 출력 해 주어야겠더라구요..

이 부분도 예외가 명확하게 문제를 표현할 수 있도록 변경했습니다!

mskangg commented on src/main/java/nextstep/session/domain/FilePathInformation.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.nio.file.Path;
+
+public class FilePathInformation {
```
### 클래스 이름을 짓는 방식
https://edu.nextstep.camp/s/E42ziiPQ/ls/tLhxXMHZ
- 클래스의 객체들이 무엇을 하는지(what he does)를 살펴본 후 기능에 기반 해서 이름을 짓는 방법
- 클래스의 객체들이 무엇인지(what he is)에 기반 해 이름을 짓는 방법

무엇인지에 대한 고민한 뒤 이름을 지어보아요!

이미지 메타데이터 + 경로를 상태로 관리하고 있군요. 🤔
`information` 은 사실 너무 범용적이고 추상적이어서 클래스나 변수 이름으로 사용하기에는 적절하지 않을 수 있습니다.  
이런 범용적인 단어는 어느 클래스에 붙여도 어색하지 않는게 단점이라 개인적으로 지양하는 편입니다.

`metadata` 가 가장 어울려보이는데, 파일 경로가 들어가 있으니..
`FileImageProperty` 가 더 어울리는것 같은데 어떠신가요?

firewoody237 replied:
음 넵! `Controller`, `Manager`, `Information` 같은 범용적인 부분은 사용하지 말아야 하는데,
제가 까먹고 있었네요!

말씀해주신 `FileImageProperty`도 좋아보이지만, 이미지 외에 다른 파일은 저장하지 않는 네이밍으로 `ImageFilePath`로 변경하였습니다!

mskangg commented on src/main/java/nextstep/payments/domain/Payment.java:
```diff
@@ -1,29 +1,5 @@
-        this.amount = amount;
-        this.createdAt = LocalDateTime.now();
-    }
+public interface Payment {
```
추상화한 이유가 있을까요? 
무료결제와 유료결제를 같은 코드로 구현하여 지금으로썬 필요가 없어보여서요 🤔

firewoody237 replied:
오 그러네요. 제가 같은 코드를 만들었군요.. ㅜㅠ
`FreePayment`는 금액이 항상 0인것 과같이 만드려는 의도 였는데, 다시 생각해보니 지금 상황에서 굳이 나누진 않아도 될 것 같습니다.
`Payment`는 롤백하였습니다!

mskangg commented on src/main/java/nextstep/session/domain/Duration.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.time.LocalDateTime;
+
+public class Duration {
```
기간을 잘 캡슐화 해주셨어요 👍

mskangg commented on src/main/java/nextstep/session/domain/Session.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface Session {
+
+    void changeStartDate(LocalDateTime startDate);
+
+    void changeEndDate(LocalDateTime endDate);
+
+    void changeCover(Cover cover);
```
setter를 다 구현해주셨네요.
강의가 모집이 됐을때 수정이 일어나면 어떻게 되는건가용??

firewoody237 replied:
빈 구멍이 있었네요..!
요소에 대한 변경은 상태가 `READY` 상태일 때만 가능하도록 코드 변경했습니다!

mskangg commented on src/main/java/nextstep/session/domain/SessionRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.Optional;
+
+public interface SessionRepository {
```
DB 레이어를 고려하여 작성하셨군요 😄
이번엔 순수 도메인만 작성하는것이 원칙이니 배제하고 구현하시는 건 어떨까용?

firewoody237 replied:
`Course` 부분이 구현되어 있어 따라 한건데,
지우고 다음미션에 다시 추가 하겠습니다~!

mskangg commented on src/main/java/nextstep/session/domain/Resolution.java:
```diff
@@ -0,0 +1,32 @@
+        this.height = height;
+    }
+
+    private void validate(int width, int height) {
```
캡슐화 잘해주셨네요

mskangg commented on src/main/java/nextstep/session/domain/Price.java:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    public boolean isFullyPaid(Payment payment) {
+        return price == payment.getPaidAmount();
```
값을 꺼내지 않고 메세지를 보내는건 어떨까요?

firewoody237 replied:
지불의 주체를 `Payment`로 바꿔 메시지를 보내도록 변경했습니다!

mskangg commented on src/test/java/nextstep/session/domain/DurationTest.java:
```diff
@@ -0,0 +1,84 @@
+
+class DurationTest {
+
+    @DisplayName("시작일자가 종료일자보다 미래면 IllegalArgumentException을 던진다.")
```
테스트명을 작성할 때 구현체를 직접 네이밍 하는것 보다

```suggestion
    @DisplayName("시작일자가 종료일자보다 미래일 수 없다.")
```

위와 같이 작성하는건 어떨까요?
개인적으로 테스트코드는 비개발자도 읽을 수 있어야한다고 생각합니다 🤔 (문서의 역할)

firewoody237 replied:
너무 멋진 관점이라고 생각합니다.
테스트 코드는 문서로서의 기능을 한다고 공부했는데, 예외와 같은 특정 명칭들이 있으면 어렵겠네요.

다른 부분들도 수정 했습니다!

mskangg reviewed (changes requested):
2단계 구현 정말 잘해주셨네요~!
코드에서 많은 고민을 하신게 단번에 느껴지네요..! 👏 
구조를 잘잡고 캡슐화 및 테스트코드도 잘 구현해주셔서 크게 피드백 드릴 부분은 보이지 않았네요 💯 
몇가지 코멘트 남겼는데 확인 부탁드립니다! 🙇

firewoody237 committed:
refactor:
- 무료강의에 대한 Capacity 위치 변경
- 모호한 명칭인 FilePathInformation -> ImageFilePath로 변경
- Payment 추상화 제거
- Generation 검증 추가

firewoody237 commented:
- 하나의 commit으로 리팩토링하면 혼동 되실 것 같아, 두개로 쪼갰습니다.
- 근데 생각해보니, 각 리팩토링 하나마다 커밋 드리는게 보시기 더 편했을 것 같습니다.. 다음번엔 꼭 유의하겠습니다!

질문이 있습니다.
3단계를 맛보기로 구현하던 중에, 도메인과 데이터베이스를 엮으려면 `getter`를 어떨 수 없이 많이 사용하는 상황이 발생했는데요,
이 부분에서 사용하는 `getter`는 어쩔 수 없는 부분일까요? 다른 방법이 있을까요?

firewoody237 committed:
refactor:
- 테스트 문서화를 위한 특정 명칭(예외명) 제거
- Session의 요소 변경에 대한 검증 추가
- Price에서의 검증을 Payment로의 메시지로 변경

firewoody237 committed:
docs: 도메인 그림 변경

mskangg commented on src/main/java/nextstep/courses/domain/Generation.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    private static void validate(int generation) {
+        if (generation < MINIMUM_GENERATION) {
+            throw new IllegalArgumentException("기수는 0이하일 수 없습니다.");
+        }
```
👍

mskangg commented on src/main/java/nextstep/payments/domain/Payment.java:
```diff
@@ -26,4 +26,12 @@ public Payment(String id, Long sessionId, Long nsUserId, Long amount) {
+        return this.amount;
+    }
+
+    public boolean isExactlyPaid(long requestAmount) {
```
👍

mskangg commented on src/main/java/nextstep/session/domain/FreeSession.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public void changeEndDate(LocalDateTime endDate) {
+        validate();
```
changeEndDate
changeCover
editSessionName

위 3개의 메서드에서 모두 validate 를 진행하고 있지만, 구체화되지 않은 메서드명 때문에 구현부를 한번더 확인해야 될것 같아요.

모두 다 변경되기전 상태를 확인하는 메서드이니 `validateChangeStatus` 와 같은 네이밍이 잘어울릴것 같습니다 😄

mskangg commented on src/main/java/nextstep/session/domain/ImageFilePath.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.nio.file.Path;
+
+public class ImageFilePath {
```
👍

mskangg commented on src/test/java/nextstep/session/domain/CapacityTest.java:
```diff
@@ -0,0 +1,73 @@
+        // given
+        Capacity capacity = Capacity.create(10);
+
+        // then
+        capacity.enroll();
+        assertThat(capacity.getEnrolled()).isEqualTo(1);
+        capacity.enroll();
+        assertThat(capacity.getEnrolled()).isEqualTo(2);
+    }
+
+    @DisplayName("수용가능 인원을 초과해서 추가할 수 없다.")
+    @Test
+    void throwCapacityExceptionWhenOverMaxCapacity() {
+        // given
+        Capacity capacity = Capacity.create(2);
+
+        // when
+        capacity.enroll();
+        capacity.enroll();
+
+        // then
+        assertThatThrownBy(() -> capacity.enroll())
+                .isInstanceOf(CapacityException.class);
+    }
+
+    @DisplayName("수강신청 인원이 최대 수용인원보다 적다면, 사용 가능하다.")
+    @Test
+    void availableWhenMasCapacityLessThanEnrolled() {
+        // given
+        Capacity capacity = Capacity.create(10);
+
+        // when
+        capacity.enroll();
+
+        // then
+        assertThat(capacity.isAvailable()).isTrue();
+    }
+
+    @DisplayName("수강신청 인원이 최대 수용인원보다 같거나 많다면, 사용 불가능하다.")
+    @Test
```
테스트 명 만으로도 문서의 역할을 충분히 하고 있네요! 💯 
좋습니다 👍

mskangg reviewed (approved):
피드백 반영 정말 잘해주셨네요! 👍 💯 
테스트코드 DisplayName 도 잘 읽히고 요구사항을 충분히 표현하고 있는것 같아요 😄
다음 단계가 기대됩니다 🚀
그럼 화이팅 🔥

mskangg commented:
@firewoody237

> 3단계를 맛보기로 구현하던 중에, 도메인과 데이터베이스를 엮으려면 getter를 어떨 수 없이 많이 사용하는 상황이 발생했는데요,
이 부분에서 사용하는 getter는 어쩔 수 없는 부분일까요? 다른 방법이 있을까요?

저도 이 고민을 많이 해봤는데, 자바빈 규약에서 get* 메서드를 사용하니 어쩔수 없는 부분도 있고,
값을 매핑하기 위해서도 필요하기 때문에 만들어 사용하는건 어쩔수 없을것 같아요 😭

말씀하신것 처럼 트레이트 오프로 get 메서드를 데이터베이스 바인딩을 위해 허용하거나, getter를 허용하지 않고 Entity 에서 데이터베이스에 바인딩 되는 VO를 만들어 리턴해주는 메서드를 작성해서 사용하곤 합니다.
하지만, entity가 vo를 의존해버리니 이또한 단점이기도 해요 😅

은탄환은 없으니 합리적인 트레이드오프를 통해 사용하시는걸 추천드립니다.
